diff -uprN linux-5.10-rc2/drivers/tty/vt/selection.c linux-5.10-rc2.new/drivers/tty/vt/selection.c
--- linux-5.10-rc2/drivers/tty/vt/selection.c	2020-11-02 07:43:51.000000000 +0900
+++ linux-5.10-rc2.new/drivers/tty/vt/selection.c	2020-11-07 19:19:52.561533230 +0900
@@ -220,6 +220,8 @@ static int vc_selection_store_chars(stru
 			}
 			obp = bp;
 		}
+		if (c > 0x80)
+			i += 2;
 	}
 	vc_sel.buf_len = bp - vc_sel.buffer;
 
diff -uprN linux-5.10-rc2/drivers/tty/vt/vt.c linux-5.10-rc2.new/drivers/tty/vt/vt.c
--- linux-5.10-rc2/drivers/tty/vt/vt.c	2020-11-02 07:43:51.000000000 +0900
+++ linux-5.10-rc2.new/drivers/tty/vt/vt.c	2020-11-07 19:20:00.425034720 +0900
@@ -297,6 +297,20 @@ static inline unsigned short *screenpos(
 	return p;
 }
 
+static inline unsigned short *screenpos_utf8(struct vc_data *vc, int offset, int viewed)
+{
+    unsigned short *p;
+
+    if (!viewed)
+        p = (unsigned short *)(vc->vc_origin + offset + vc->vc_screenbuf_size);
+    else if (!vc->vc_sw->con_screen_pos)
+        p = (unsigned short *)(vc->vc_visible_origin + offset + vc->vc_screenbuf_size);
+    else
+        p = vc->vc_sw->con_screen_pos(vc, -offset - 1);
+    return p;
+}
+
+
 /* Called  from the keyboard irq path.. */
 static inline void scrolldelta(int lines)
 {
@@ -639,6 +653,11 @@ static void con_scroll(struct vc_data *v
 	}
 	scr_memmovew(d, s, (b - t - nr) * vc->vc_size_row);
 	scr_memsetw(clear, vc->vc_video_erase_char, vc->vc_size_row * nr);
+
+    d += (vc->vc_screenbuf_size >> 1);
+    s += (vc->vc_screenbuf_size >> 1);
+    scr_memmovew(d, s, (b - t - nr) * vc->vc_size_row);
+    scr_memsetw(d + (b - t - nr) * vc->vc_cols, 0, vc->vc_size_row * nr);
 }
 
 static void do_update_region(struct vc_data *vc, unsigned long start, int count)
@@ -807,6 +826,8 @@ void complement_pos(struct vc_data *vc,
 	static int old_offset = -1;
 	static unsigned short old;
 	static unsigned short oldx, oldy;
+    static unsigned short *p_ext = NULL;
+    static unsigned short old_ext = 0;
 
 	WARN_CONSOLE_UNLOCKED();
 
@@ -814,7 +835,7 @@ void complement_pos(struct vc_data *vc,
 	    old_offset < vc->vc_screenbuf_size) {
 		scr_writew(old, screenpos(vc, old_offset, true));
 		if (con_should_update(vc))
-			vc->vc_sw->con_putc(vc, old, oldy, oldx);
+            vc->vc_sw->con_putc(vc, (old_ext << 16)|old, oldy, oldx);
 		notify_update(vc);
 	}
 
@@ -831,7 +852,7 @@ void complement_pos(struct vc_data *vc,
 		if (con_should_update(vc)) {
 			oldx = (offset >> 1) % vc->vc_cols;
 			oldy = (offset >> 1) / vc->vc_cols;
-			vc->vc_sw->con_putc(vc, new, oldy, oldx);
+            vc->vc_sw->con_putc(vc, (old_ext << 16)|new, oldy, oldx);
 		}
 		notify_update(vc);
 	}
@@ -1145,7 +1166,7 @@ int vc_allocate(unsigned int currcons)	/
 	    vc->vc_screenbuf_size > KMALLOC_MAX_SIZE || !vc->vc_screenbuf_size)
 		goto err_free;
 	err = -ENOMEM;
-	vc->vc_screenbuf = kzalloc(vc->vc_screenbuf_size, GFP_KERNEL);
+    vc->vc_screenbuf = kzalloc(vc->vc_screenbuf_size * 2, GFP_KERNEL);
 	if (!vc->vc_screenbuf)
 		goto err_free;
 
@@ -1198,7 +1219,8 @@ static int vc_do_resize(struct tty_struc
 {
 	unsigned long old_origin, new_origin, new_scr_end, rlth, rrem, err = 0;
 	unsigned long end;
-	unsigned int old_rows, old_row_size, first_copied_row;
+    /* unsigned int old_rows, old_row_size, first_copied_row; */
+    unsigned int old_rows, old_row_size, old_screen_size, first_copied_row;
 	unsigned int new_cols, new_rows, new_row_size, new_screen_size;
 	unsigned int user;
 	unsigned short *oldscreen, *newscreen;
@@ -1219,13 +1241,14 @@ static int vc_do_resize(struct tty_struc
 	new_rows = (lines ? lines : vc->vc_rows);
 	new_row_size = new_cols << 1;
 	new_screen_size = new_row_size * new_rows;
+    old_screen_size = old_rows * old_row_size;
 
 	if (new_cols == vc->vc_cols && new_rows == vc->vc_rows)
 		return 0;
 
 	if (new_screen_size > KMALLOC_MAX_SIZE || !new_screen_size)
 		return -EINVAL;
-	newscreen = kzalloc(new_screen_size, GFP_USER);
+    newscreen = kzalloc(new_screen_size * 2, GFP_USER);
 	if (!newscreen)
 		return -ENOMEM;
 
@@ -1290,15 +1313,23 @@ static int vc_do_resize(struct tty_struc
 	while (old_origin < end) {
 		scr_memcpyw((unsigned short *) new_origin,
 			    (unsigned short *) old_origin, rlth);
-		if (rrem)
+        scr_memcpyw((unsigned short *) new_origin + (new_screen_size >> 1),
+                (unsigned short *) old_origin + (old_screen_size >> 1), rlth);
+        if (rrem){
 			scr_memsetw((void *)(new_origin + rlth),
 				    vc->vc_video_erase_char, rrem);
+            scr_memsetw((void *)(new_origin + rlth + (new_screen_size)),
+                    vc->vc_video_erase_char, rrem);
+        }
 		old_origin += old_row_size;
 		new_origin += new_row_size;
 	}
-	if (new_scr_end > new_origin)
+    if (new_scr_end > new_origin){
 		scr_memsetw((void *)new_origin, vc->vc_video_erase_char,
 			    new_scr_end - new_origin);
+        scr_memsetw((void *)new_origin + (new_screen_size), vc->vc_video_erase_char,
+                new_scr_end - new_origin);
+    }
 	oldscreen = vc->vc_screenbuf;
 	vc->vc_screenbuf = newscreen;
 	vc->vc_screenbuf_size = new_screen_size;
@@ -2814,12 +2845,28 @@ static int vc_con_write_normal(struct vc
 			insert_char(vc, 1);
 		vc_uniscr_putc(vc, next_c);
 
-		if (himask)
-			tc = ((tc & 0x100) ? himask : 0) |
-			      (tc &  0xff);
-		tc |= (vc_attr << 8) & ~himask;
-
-		scr_writew(tc, (u16 *)vc->vc_pos);
+        if(is_double_width(c) && width==2)
+        {
+            tc = 0xFF;
+            scr_writew(himask ?
+                    ((vc_attr << 8) & ~himask) + ((tc & 0x100) ? himask : 0) + (tc & 0xff) :
+                    (vc_attr << 8) + tc,
+                    (u16 *) vc->vc_pos);
+            scr_writew(c, (u16 *) vc->vc_pos + (vc->vc_screenbuf_size >> 1));
+        }else if(is_double_width(c) && width==1){
+            tc = 0xFE;
+            scr_writew(himask ?
+                    ((vc_attr << 8) & ~himask) + ((tc & 0x100) ? himask : 0) + (tc & 0xff) :
+                    (vc_attr << 8) + tc,
+                    (u16 *) vc->vc_pos);
+            scr_writew(c, (u16 *) vc->vc_pos + (vc->vc_screenbuf_size >> 1));
+        }else{
+            scr_writew(himask ?
+                    ((vc_attr << 8) & ~himask) + ((tc & 0x100) ? himask : 0) + (tc & 0xff) :
+                    (vc_attr << 8) + tc,
+                    (u16 *) vc->vc_pos);
+            scr_writew(c, (u16 *) vc->vc_pos + (vc->vc_screenbuf_size >> 1));
+        }
 
 		if (con_should_update(vc) && draw->x < 0) {
 			draw->x = vc->state.x;
@@ -2896,6 +2943,8 @@ static int do_con_write(struct tty_struc
 rescan_last_byte:
 		c = orig;
 		rescan = false;
+        vc->vc_utf = 1;
+        vc->vc_disp_ctrl = 0;
 
 		tc = vc_translate(vc, &c, &rescan);
 		if (tc == -1)
@@ -3490,7 +3539,7 @@ static int __init con_init(void)
 		tty_port_init(&vc->port);
 		visual_init(vc, currcons, 1);
 		/* Assuming vc->vc_{cols,rows,screenbuf_size} are sane here. */
-		vc->vc_screenbuf = kzalloc(vc->vc_screenbuf_size, GFP_NOWAIT);
+        vc->vc_screenbuf = kzalloc(vc->vc_screenbuf_size * 2, GFP_NOWAIT);
 		vc_init(vc, vc->vc_rows, vc->vc_cols,
 			currcons || !vc->vc_sw->con_save_screen);
 	}
@@ -4727,6 +4776,11 @@ static int con_font_copy(struct vc_data
 
 int con_font_op(struct vc_data *vc, struct console_font_op *op)
 {
+    /* IF WE ENABLE CJK font, I won't let you guys set console font any way until
+       we found a way to set console font that is unicode capable
+       */
+    if(vc->vc_font.charcount > op->charcount)
+        return 0;
 	switch (op->op) {
 	case KD_FONT_OP_SET:
 		return con_font_set(vc, op);
@@ -4750,10 +4804,16 @@ u16 screen_glyph(const struct vc_data *v
 	u16 w = scr_readw(screenpos(vc, offset, true));
 	u16 c = w & 0xff;
 
+    u16 c_utf8 = scr_readw(screenpos_utf8(vc, offset, 1));
+
+    if ( (c == 0xff || c == 0xfe) && c_utf8 != 0){
+        return c_utf8;
+    }else{
 	if (w & vc->vc_hi_font_mask)
 		c |= 0x100;
 	return c;
 }
+}
 EXPORT_SYMBOL_GPL(screen_glyph);
 
 u32 screen_glyph_unicode(const struct vc_data *vc, int n)
diff -uprN linux-5.10-rc2/drivers/video/console/Kconfig linux-5.10-rc2.new/drivers/video/console/Kconfig
--- linux-5.10-rc2/drivers/video/console/Kconfig	2020-11-02 07:43:51.000000000 +0900
+++ linux-5.10-rc2.new/drivers/video/console/Kconfig	2020-11-07 19:19:52.584867063 +0900
@@ -97,6 +97,7 @@ config FRAMEBUFFER_CONSOLE_DETECT_PRIMAR
 config FRAMEBUFFER_CONSOLE_ROTATION
        bool "Framebuffer Console Rotation"
        depends on FRAMEBUFFER_CONSOLE
+       depends on !FONT_16x16_CJK
        help
 	 Enable display rotation for the framebuffer console.  This is done
 	 in software and may be significantly slower than a normally oriented
diff -uprN linux-5.10-rc2/drivers/video/console/Makefile linux-5.10-rc2.new/drivers/video/console/Makefile
--- linux-5.10-rc2/drivers/video/console/Makefile	2020-11-02 07:43:51.000000000 +0900
+++ linux-5.10-rc2.new/drivers/video/console/Makefile	2020-11-07 19:19:52.588200467 +0900
@@ -3,6 +3,7 @@
 # 5 Aug 1999, James Simmons, <mailto:jsimmons@users.sf.net>
 # Rewritten to use lists instead of if-statements.
 
+# Font handling
 obj-$(CONFIG_DUMMY_CONSOLE)       += dummycon.o
 obj-$(CONFIG_SGI_NEWPORT_CONSOLE) += newport_con.o
 obj-$(CONFIG_STI_CONSOLE)         += sticon.o sticore.o
diff -uprN linux-5.10-rc2/drivers/video/fbdev/core/bitblit.c linux-5.10-rc2.new/drivers/video/fbdev/core/bitblit.c
--- linux-5.10-rc2/drivers/video/fbdev/core/bitblit.c	2020-11-02 07:43:51.000000000 +0900
+++ linux-5.10-rc2.new/drivers/video/fbdev/core/bitblit.c	2020-11-07 19:19:52.611534300 +0900
@@ -10,6 +10,7 @@
  *  more details.
  */
 
+#include <linux/font.h>
 #include <linux/module.h>
 #include <linux/slab.h>
 #include <linux/string.h>
@@ -43,6 +44,23 @@ static void update_attr(u8 *dst, u8 *src
 	}
 }
 
+static inline u16 utf8_pos(struct vc_data *vc, const unsigned short *utf8)
+{
+	unsigned long p = (long)utf8;
+	if (p >= vc->vc_origin && p < vc->vc_scr_end) {
+		return scr_readw((unsigned short *)(p + vc->vc_screenbuf_size));
+/*
+	} else if (vc->vc_num == fg_console && fbcon_is_softback(utf8)){
+		return scr_readw((unsigned short *)(p + fbcon_softback_size));
+*/
+	} else {
+		u16 extra_c;
+		int c = *(int*)utf8;
+		extra_c = (c >> 16 ) & 0x0000ffff;
+		return extra_c;
+	}
+}
+
 static void bit_bmove(struct vc_data *vc, struct fb_info *info, int sy,
 		      int sx, int dy, int dx, int height, int width)
 {
@@ -74,6 +92,34 @@ static void bit_clear(struct vc_data *vc
 	info->fbops->fb_fillrect(info, &region);
 }
 
+u8 * font_bits(struct vc_data *vc, const u16 *s,u32 cellsize,
+					u16 charmask)
+{
+	u32	utf8_c;
+
+	u8 *src = vc->vc_font.data + (scr_readw(s)&
+				  charmask)*cellsize;
+
+	utf8_c = utf8_pos(vc, s);
+
+	if( utf8_c <= vc->vc_font.charcount)
+	{
+		/*
+		 * decide left-half char or right-half char.
+		 * Since non-English chars may double weight
+		 */
+		switch (scr_readw(s) & charmask) {
+			case 0xff:
+				src = vc->vc_font.data + (utf8_c * cellsize *2 );
+				break;
+			case 0xfe:
+				src = vc->vc_font.data + (utf8_c * cellsize *2 + cellsize);
+				break;
+		}
+	}
+	return src;
+}
+
 static inline void bit_putcs_aligned(struct vc_data *vc, struct fb_info *info,
 				     const u16 *s, u32 attr, u32 cnt,
 				     u32 d_pitch, u32 s_pitch, u32 cellsize,
@@ -84,14 +130,12 @@ static inline void bit_putcs_aligned(str
 	u8 *src;
 
 	while (cnt--) {
-		src = vc->vc_font.data + (scr_readw(s++)&
-					  charmask)*cellsize;
+		src = font_bits(vc,s++,cellsize,charmask);
 
 		if (attr) {
 			update_attr(buf, src, attr, vc);
 			src = buf;
 		}
-
 		if (likely(idx == 1))
 			__fb_pad_aligned_buffer(dst, d_pitch, src, idx,
 						image->height);
@@ -119,14 +163,11 @@ static inline void bit_putcs_unaligned(s
 	u8 *src;
 
 	while (cnt--) {
-		src = vc->vc_font.data + (scr_readw(s++)&
-					  charmask)*cellsize;
-
+		src = font_bits(vc,s++,cellsize,charmask);
 		if (attr) {
 			update_attr(buf, src, attr, vc);
 			src = buf;
 		}
-
 		fb_pad_unaligned_buffer(dst, d_pitch, src, idx,
 					image->height, shift_high,
 					shift_low, mod);
@@ -245,11 +286,13 @@ static void bit_cursor(struct vc_data *v
 	int err = 1;
 	char *src;
 
+	int cellsize = DIV_ROUND_UP(vc->vc_font.width,8) * vc->vc_font.height;
+
 	cursor.set = 0;
 
  	c = scr_readw((u16 *) vc->vc_pos);
 	attribute = get_attribute(info, c);
-	src = vc->vc_font.data + ((c & charmask) * (w * vc->vc_font.height));
+	src = font_bits(vc,(u16*)vc->vc_pos,cellsize,charmask);
 
 	if (ops->cursor_state.image.data != src ||
 	    ops->cursor_reset) {
diff -uprN linux-5.10-rc2/drivers/video/fbdev/core/fbcon.c linux-5.10-rc2.new/drivers/video/fbdev/core/fbcon.c
--- linux-5.10-rc2/drivers/video/fbdev/core/fbcon.c	2020-11-02 07:43:51.000000000 +0900
+++ linux-5.10-rc2.new/drivers/video/fbdev/core/fbcon.c	2020-11-07 19:19:52.614867705 +0900
@@ -206,6 +206,16 @@ static void fbcon_start(void);
 static void fbcon_exit(void);
 static struct device *fbcon_device;
 
+/*
+int fbcon_is_softback(const unsigned short *str)
+{
+	unsigned long p = (long)str;
+	if(p >= softback_buf && p <softback_end)
+		return 1;
+	return 0;
+}
+*/
+
 #ifdef CONFIG_FRAMEBUFFER_CONSOLE_ROTATION
 static inline void fbcon_set_rotation(struct fb_info *info)
 {
@@ -1004,7 +1014,8 @@ static const char *fbcon_startup(void)
 		vc->vc_font.width = font->width;
 		vc->vc_font.height = font->height;
 		vc->vc_font.data = (void *)(p->fontdata = font->data);
-		vc->vc_font.charcount = 256; /* FIXME  Need to support more fonts */
+		// read charcount from font_desc, yep, finally fixed
+		vc->vc_font.charcount = font->charcount;
 	} else {
 		p->fontdata = vc->vc_font.data;
 	}
@@ -1068,6 +1079,7 @@ static void fbcon_init(struct vc_data *v
 						    fvc->vc_font.data);
 			vc->vc_font.width = fvc->vc_font.width;
 			vc->vc_font.height = fvc->vc_font.height;
+			vc->vc_font.charcount = fvc->vc_font.charcount;
 			p->userfont = t->userfont;
 
 			if (p->userfont)
@@ -1083,8 +1095,8 @@ static void fbcon_init(struct vc_data *v
 			vc->vc_font.width = font->width;
 			vc->vc_font.height = font->height;
 			vc->vc_font.data = (void *)(p->fontdata = font->data);
-			vc->vc_font.charcount = 256; /* FIXME  Need to
-							support more fonts */
+			// read charcount from font , finnaly fixed
+			vc->vc_font.charcount = font->charcount;
 		}
 	}
 
@@ -1310,10 +1322,7 @@ static void fbcon_putcs(struct vc_data *
 
 static void fbcon_putc(struct vc_data *vc, int c, int ypos, int xpos)
 {
-	unsigned short chr;
-
-	scr_writew(c, &chr);
-	fbcon_putcs(vc, &chr, 1, ypos, xpos);
+	fbcon_putcs(vc, (unsigned short *)&c, 1, ypos, xpos);
 }
 
 static void fbcon_clear_margins(struct vc_data *vc, int bottom_only)
@@ -1378,6 +1387,7 @@ static void fbcon_set_disp(struct fb_inf
 		vc->vc_font.data = (void *)(p->fontdata = t->fontdata);
 		vc->vc_font.width = (*default_mode)->vc_font.width;
 		vc->vc_font.height = (*default_mode)->vc_font.height;
+		vc->vc_font.charcount = (*default_mode)->vc_font.charcount;
 		p->userfont = t->userfont;
 		if (p->userfont)
 			REFCOUNT(p->fontdata)++;
@@ -1638,6 +1648,7 @@ static void fbcon_redraw_blit(struct vc_
 static void fbcon_redraw(struct vc_data *vc, struct fbcon_display *p,
 			 int line, int count, int offset)
 {
+    u16 charmask = vc->vc_hi_font_mask ? 0x1ff : 0xff;
 	unsigned short *d = (unsigned short *)
 	    (vc->vc_origin + vc->vc_size_row * line);
 	unsigned short *s = d + offset;
@@ -1660,6 +1671,8 @@ static void fbcon_redraw(struct vc_data
 					start = s;
 				}
 			}
+            if( ((scr_readw(s) & charmask) == 0xff || (scr_readw(s) & charmask) == 0xfe) && scr_readw(s + (vc->vc_screenbuf_size >> 1)) != 0){
+            }else{
 			if (c == scr_readw(d)) {
 				if (s > start) {
 					fbcon_putcs(vc, start, s - start,
@@ -1671,7 +1684,9 @@ static void fbcon_redraw(struct vc_data
 					start++;
 				}
 			}
+            }
 			scr_writew(c, d);
+            scr_writew(scr_readw(s + (vc->vc_screenbuf_size >> 1)), d + (vc->vc_screenbuf_size >> 1));
 			console_conditional_schedule();
 			s++;
 			d++;
@@ -2289,7 +2304,6 @@ static int fbcon_get_font(struct vc_data
 
 	font->width = vc->vc_font.width;
 	font->height = vc->vc_font.height;
-	font->charcount = vc->vc_hi_font_mask ? 512 : 256;
 	if (!font->data)
 		return 0;
 
diff -uprN linux-5.10-rc2/drivers/video/fbdev/core/fbcon.h linux-5.10-rc2.new/drivers/video/fbdev/core/fbcon.h
--- linux-5.10-rc2/drivers/video/fbdev/core/fbcon.h	2020-11-02 07:43:51.000000000 +0900
+++ linux-5.10-rc2.new/drivers/video/fbdev/core/fbcon.h	2020-11-07 19:19:52.614867705 +0900
@@ -255,4 +255,7 @@ extern void fbcon_set_rotate(struct fbco
 #define fbcon_set_rotate(x) do {} while(0)
 #endif /* CONFIG_FRAMEBUFFER_CONSOLE_ROTATION */
 
+extern int fbcon_softback_size;
+extern int fbcon_is_softback(const unsigned short *str);
+
 #endif /* _VIDEO_FBCON_H */
diff -uprN linux-5.10-rc2/drivers/video/fbdev/core/fbcon_ccw.c linux-5.10-rc2.new/drivers/video/fbdev/core/fbcon_ccw.c
--- linux-5.10-rc2/drivers/video/fbdev/core/fbcon_ccw.c	2020-11-02 07:43:51.000000000 +0900
+++ linux-5.10-rc2.new/drivers/video/fbdev/core/fbcon_ccw.c	2020-11-07 19:19:52.614867705 +0900
@@ -18,6 +18,8 @@
 #include "fbcon.h"
 #include "fbcon_rotate.h"
 
+extern u8 * font_bits(struct vc_data *vc, const u16 *s,u32 cellsize,
+		u16 charmask);
 /*
  * Rotation 270 degrees
  */
@@ -105,13 +107,23 @@ static inline void ccw_putcs_aligned(str
 	u32 idx = (vc->vc_font.height + 7) >> 3;
 	u8 *src;
 
-	while (cnt--) {
-		src = ops->fontbuffer + (scr_readw(s--) & charmask)*cellsize;
 
+	while (cnt--) {
+		if(((scr_readw(s) & charmask) == 0xff || (scr_readw(s) & charmask) == 0xfe )){
+			char dst[16];
+			src = font_bits(vc,s,cellsize,charmask);
+			memset(dst, 0, 16);
+			rotate_ccw(src, dst, vc->vc_font.width,
+				  vc->vc_font.height);
+			src = dst;
+		}else{
+			src = ops->fontbuffer + (scr_readw(s) & charmask)*cellsize;
+		}
 		if (attr) {
 			ccw_update_attr(buf, src, attr, vc);
 			src = buf;
 		}
+		s--;
 
 		if (likely(idx == 1))
 			__fb_pad_aligned_buffer(dst, d_pitch, src, idx,
@@ -230,6 +242,7 @@ static void ccw_cursor(struct vc_data *v
 	int err = 1, dx, dy;
 	char *src;
 	u32 vyres = GETVYRES(ops->p->scrollmode, info);
+	int cellsize = DIV_ROUND_UP(vc->vc_font.width,8) * vc->vc_font.height;
 
 	if (!ops->fontbuffer)
 		return;
@@ -238,7 +251,17 @@ static void ccw_cursor(struct vc_data *v
 
  	c = scr_readw((u16 *) vc->vc_pos);
 	attribute = get_attribute(info, c);
+
+	if(((c&charmask) == 0xff || (c & charmask) == 0xfe)){
+		char dst[16];
+		src = font_bits(vc,(const u16*)&c,cellsize,charmask);
+		memset(dst, 0, 16);
+		rotate_ccw(src, dst, vc->vc_font.width,
+			  vc->vc_font.height);
+		src = dst;
+	}else{
 	src = ops->fontbuffer + ((c & charmask) * (w * vc->vc_font.width));
+	}
 
 	if (ops->cursor_state.image.data != src ||
 	    ops->cursor_reset) {
diff -uprN linux-5.10-rc2/drivers/video/fbdev/core/fbcon_cw.c linux-5.10-rc2.new/drivers/video/fbdev/core/fbcon_cw.c
--- linux-5.10-rc2/drivers/video/fbdev/core/fbcon_cw.c	2020-11-02 07:43:51.000000000 +0900
+++ linux-5.10-rc2.new/drivers/video/fbdev/core/fbcon_cw.c	2020-11-07 19:19:52.614867705 +0900
@@ -18,6 +18,8 @@
 #include "fbcon.h"
 #include "fbcon_rotate.h"
 
+extern u8 * font_bits(struct vc_data *vc, const u16 *s,u32 cellsize,
+		u16 charmask);
 /*
  * Rotation 90 degrees
  */
@@ -89,14 +91,23 @@ static inline void cw_putcs_aligned(stru
 	u16 charmask = vc->vc_hi_font_mask ? 0x1ff : 0xff;
 	u32 idx = (vc->vc_font.height + 7) >> 3;
 	u8 *src;
-
 	while (cnt--) {
-		src = ops->fontbuffer + (scr_readw(s++) & charmask)*cellsize;
 
+		if(((scr_readw(s) & charmask) == 0xff || (scr_readw(s) & charmask) == 0xfe )){
+			char dst[16];
+			src = font_bits(vc,s,cellsize,charmask);
+			memset(dst, 0, 16);
+			rotate_cw(src, dst, vc->vc_font.width,
+				  vc->vc_font.height);
+			src = dst;
+		}else{
+			src = ops->fontbuffer + (scr_readw(s) & charmask)*cellsize;
+		}
 		if (attr) {
 			cw_update_attr(buf, src, attr, vc);
 			src = buf;
 		}
+		s++;
 
 		if (likely(idx == 1))
 			__fb_pad_aligned_buffer(dst, d_pitch, src, idx,
@@ -213,6 +224,7 @@ static void cw_cursor(struct vc_data *vc
 	int err = 1, dx, dy;
 	char *src;
 	u32 vxres = GETVXRES(ops->p->scrollmode, info);
+	int cellsize = DIV_ROUND_UP(vc->vc_font.width,8) * vc->vc_font.height;
 
 	if (!ops->fontbuffer)
 		return;
@@ -221,7 +233,16 @@ static void cw_cursor(struct vc_data *vc
 
  	c = scr_readw((u16 *) vc->vc_pos);
 	attribute = get_attribute(info, c);
+	if(((c&charmask) == 0xff || (c & charmask) == 0xfe)){
+		char dst[16];
+		src = font_bits(vc,(const u16*)&c,cellsize,charmask);
+		memset(dst, 0, 16);
+		rotate_cw(src, dst, vc->vc_font.width,
+			  vc->vc_font.height);
+		src = dst;
+	}else{
 	src = ops->fontbuffer + ((c & charmask) * (w * vc->vc_font.width));
+	}
 
 	if (ops->cursor_state.image.data != src ||
 	    ops->cursor_reset) {
diff -uprN linux-5.10-rc2/drivers/video/fbdev/core/fbcon_ud.c linux-5.10-rc2.new/drivers/video/fbdev/core/fbcon_ud.c
--- linux-5.10-rc2/drivers/video/fbdev/core/fbcon_ud.c	2020-11-02 07:43:51.000000000 +0900
+++ linux-5.10-rc2.new/drivers/video/fbdev/core/fbcon_ud.c	2020-11-07 19:19:52.618201109 +0900
@@ -18,6 +18,8 @@
 #include "fbcon.h"
 #include "fbcon_rotate.h"
 
+extern u8 * font_bits(struct vc_data *vc, const u16 *s,u32 cellsize,
+		u16 charmask);
 /*
  * Rotation 180 degrees
  */
@@ -93,12 +95,22 @@ static inline void ud_putcs_aligned(stru
 	u8 *src;
 
 	while (cnt--) {
-		src = ops->fontbuffer + (scr_readw(s--) & charmask)*cellsize;
 
+		if(((scr_readw(s) & charmask) == 0xff || (scr_readw(s) & charmask) == 0xfe ) ){
+			char dst[16];
+			src = font_bits(vc,s,cellsize,charmask);
+			memset(dst, 0, 16);
+			rotate_ud(src, dst, vc->vc_font.width,
+				  vc->vc_font.height);
+			src = dst;
+		}else{
+			src = ops->fontbuffer + (scr_readw(s) & charmask)*cellsize;
+		}
 		if (attr) {
 			ud_update_attr(buf, src, attr, vc);
 			src = buf;
 		}
+		s--;
 
 		if (likely(idx == 1))
 			__fb_pad_aligned_buffer(dst, d_pitch, src, idx,
@@ -128,12 +140,21 @@ static inline void ud_putcs_unaligned(st
 	u8 *src;
 
 	while (cnt--) {
-		src = ops->fontbuffer + (scr_readw(s--) & charmask)*cellsize;
-
+		if(((scr_readw(s) & charmask) == 0xff || (scr_readw(s) & charmask) == 0xfe )){
+			char dst[16];
+			src = font_bits(vc,s,cellsize,charmask);
+			memset(dst, 0, 16);
+			rotate_ud(src, dst, vc->vc_font.width,
+				  vc->vc_font.height);
+			src = dst;
+		}else{
+			src = ops->fontbuffer + (scr_readw(s) & charmask)*cellsize;
+		}
 		if (attr) {
 			ud_update_attr(buf, src, attr, vc);
 			src = buf;
 		}
+		s--;
 
 		fb_pad_unaligned_buffer(dst, d_pitch, src, idx,
 					image->height, shift_high,
@@ -261,6 +282,7 @@ static void ud_cursor(struct vc_data *vc
 	char *src;
 	u32 vyres = GETVYRES(ops->p->scrollmode, info);
 	u32 vxres = GETVXRES(ops->p->scrollmode, info);
+	int cellsize = DIV_ROUND_UP(vc->vc_font.width,8) * vc->vc_font.height;
 
 	if (!ops->fontbuffer)
 		return;
@@ -269,7 +291,18 @@ static void ud_cursor(struct vc_data *vc
 
  	c = scr_readw((u16 *) vc->vc_pos);
 	attribute = get_attribute(info, c);
+	if(((c&charmask) == 0xff || (c & charmask) == 0xfe) ){
+		char dst[16];
+
+		src = font_bits(vc,(u16*)&c,cellsize,charmask);
+
+		memset(dst, 0, 16);
+		rotate_ud(src, dst, vc->vc_font.width,
+			  vc->vc_font.height);
+		src = dst;
+	}else{
 	src = ops->fontbuffer + ((c & charmask) * (w * vc->vc_font.height));
+	}
 
 	if (ops->cursor_state.image.data != src ||
 	    ops->cursor_reset) {
diff -uprN linux-5.10-rc2/include/linux/font.h linux-5.10-rc2.new/include/linux/font.h
--- linux-5.10-rc2/include/linux/font.h	2020-11-02 07:43:51.000000000 +0900
+++ linux-5.10-rc2.new/include/linux/font.h	2020-11-07 19:21:24.946808837 +0900
@@ -19,6 +19,7 @@ struct font_desc {
     int width, height;
     const void *data;
     int pref;
+    int charcount;
 };
 
 #define VGA8x8_IDX	0
@@ -34,6 +35,7 @@ struct font_desc {
 #define FONT6x10_IDX	10
 #define TER16x32_IDX	11
 #define FONT6x8_IDX	12
+#define FONT_16x16_CJK_IDX 13
 
 extern const struct font_desc	font_vga_8x8,
 			font_vga_8x16,
@@ -47,7 +49,8 @@ extern const struct font_desc	font_vga_8
 			font_mini_4x6,
 			font_6x10,
 			font_ter_16x32,
-			font_6x8;
+			font_6x8,
+            font_16x16_cjk;
 
 /* Find a font with a specific name */
 
diff -uprN linux-5.10-rc2/lib/fonts/Kconfig linux-5.10-rc2.new/lib/fonts/Kconfig
--- linux-5.10-rc2/lib/fonts/Kconfig	2020-11-02 07:43:51.000000000 +0900
+++ linux-5.10-rc2.new/lib/fonts/Kconfig	2020-11-07 19:21:58.557511605 +0900
@@ -125,8 +125,27 @@ config FONT_6x8
 	help
 	  This font is useful for small displays (OLED).
 
+config FONT_10x18
+	bool "console 10x18 font (not supported by all drivers)" if FONTS
+	depends on FRAMEBUFFER_CONSOLE
+	help
+	  This is a high resolution console font for machines with very
+	  big letters. It fits between the sun 12x22 and the normal 8x16 font.
+	  If other fonts are too big or too small for you, say Y, otherwise say N.
+
+config FONT_16x16_CJK
+	bool "console 16x16 CJK font ( cover BMP )" if FONTS
+	depends on FRAMEBUFFER_CONSOLE
+	depends on FB
+	depends on VT_CONSOLE
+	select CONSOLE_TRANSLATIONS
+	help
+	  This is a high resolution console font that covers UNICODE bmp,
+	  mostly is CJK font. If you want full UNICODE support on the VT console, say Y.
+
 config FONT_AUTOSELECT
 	def_bool y
+	depends on FRAMEBUFFER_CONSOLE || SGI_NEWPORT_CONSOLE || STI_CONSOLE || USB_SISUSBVGA_CON
 	depends on !FONT_8x8
 	depends on !FONT_6x11
 	depends on !FONT_7x14
@@ -139,6 +158,6 @@ config FONT_AUTOSELECT
 	depends on !FONT_10x18
 	depends on !FONT_TER16x32
 	depends on !FONT_6x8
-	select FONT_8x16
+	select FONT_16x16_CJK
 
 endif # FONT_SUPPORT
diff -uprN linux-5.10-rc2/lib/fonts/Makefile linux-5.10-rc2.new/lib/fonts/Makefile
--- linux-5.10-rc2/lib/fonts/Makefile	2020-11-02 07:43:51.000000000 +0900
+++ linux-5.10-rc2.new/lib/fonts/Makefile	2020-11-07 19:22:14.981188346 +0900
@@ -16,6 +16,7 @@ font-objs-$(CONFIG_FONT_MINI_4x6)  += fo
 font-objs-$(CONFIG_FONT_6x10)      += font_6x10.o
 font-objs-$(CONFIG_FONT_TER16x32)  += font_ter16x32.o
 font-objs-$(CONFIG_FONT_6x8)       += font_6x8.o
+font-objs-$(CONFIG_FONT_16x16_CJK) += font_16x16_cjk.o
 
 font-objs += $(font-objs-y)
 
diff -uprN linux-5.10-rc2/lib/fonts/font_10x18.c linux-5.10-rc2.new/lib/fonts/font_10x18.c
--- linux-5.10-rc2/lib/fonts/font_10x18.c	2020-11-02 07:43:51.000000000 +0900
+++ linux-5.10-rc2.new/lib/fonts/font_10x18.c	2020-11-07 19:19:52.688202606 +0900
@@ -5143,4 +5143,5 @@ const struct font_desc font_10x18 = {
 #else
 	.pref	= -1,
 #endif
+	.charcount = 255,
 };
diff -uprN linux-5.10-rc2/lib/fonts/font_16x16_cjk.c linux-5.10-rc2.new/lib/fonts/font_16x16_cjk.c
--- linux-5.10-rc2/lib/fonts/font_16x16_cjk.c	1970-01-01 09:00:00.000000000 +0900
+++ linux-5.10-rc2.new/lib/fonts/font_16x16_cjk.c	2020-11-07 19:19:52.688202606 +0900
@@ -0,0 +1,24 @@
+/*************************************************/
+/*                                               */
+/*       Font file modified from 		 */
+/*  http://blog.chinaunix.net/u/13265/showart.php?id=1008020       */
+/*  microcaicai@gmail modifiy it to use in-kernel*/
+/*  font solution			         */
+/*                                               */
+/*************************************************/
+
+#include <linux/font.h>
+
+static const unsigned char fontdata_16x16[] = {
+	#include "font_cjk.h"
+};
+
+const struct font_desc font_16x16_cjk = {
+	.idx	= FONT_16x16_CJK_IDX,
+	.name	= "VGA_CJK",
+	.width	= 8, // have to do this to make curser appear 8dots length
+	.height	= 16,
+	.data	= fontdata_16x16,
+	.pref	= 10, // make it big enough to be selected
+	.charcount = 65535,
+};
diff -uprN linux-5.10-rc2/lib/fonts/font_6x11.c linux-5.10-rc2.new/lib/fonts/font_6x11.c
--- linux-5.10-rc2/lib/fonts/font_6x11.c	2020-11-02 07:43:51.000000000 +0900
+++ linux-5.10-rc2.new/lib/fonts/font_6x11.c	2020-11-07 19:19:52.688202606 +0900
@@ -3349,4 +3349,5 @@ const struct font_desc font_vga_6x11 = {
 	.data	= fontdata_6x11.data,
 	/* Try avoiding this font if possible unless on MAC */
 	.pref	= -2000,
+	.charcount = 255,
 };
diff -uprN linux-5.10-rc2/lib/fonts/font_7x14.c linux-5.10-rc2.new/lib/fonts/font_7x14.c
--- linux-5.10-rc2/lib/fonts/font_7x14.c	2020-11-02 07:43:51.000000000 +0900
+++ linux-5.10-rc2.new/lib/fonts/font_7x14.c	2020-11-07 19:19:52.691536011 +0900
@@ -4115,4 +4115,5 @@ const struct font_desc font_7x14 = {
 	.height	= 14,
 	.data	= fontdata_7x14.data,
 	.pref	= 0,
+	.charcount = 255,
 };
diff -uprN linux-5.10-rc2/lib/fonts/font_8x16.c linux-5.10-rc2.new/lib/fonts/font_8x16.c
--- linux-5.10-rc2/lib/fonts/font_8x16.c	2020-11-02 07:43:51.000000000 +0900
+++ linux-5.10-rc2.new/lib/fonts/font_8x16.c	2020-11-07 19:19:52.694869415 +0900
@@ -4629,5 +4629,6 @@ const struct font_desc font_vga_8x16 = {
 	.height	= 16,
 	.data	= fontdata_8x16.data,
 	.pref	= 0,
+	.charcount = 255,
 };
 EXPORT_SYMBOL(font_vga_8x16);
diff -uprN linux-5.10-rc2/lib/fonts/font_8x8.c linux-5.10-rc2.new/lib/fonts/font_8x8.c
--- linux-5.10-rc2/lib/fonts/font_8x8.c	2020-11-02 07:43:51.000000000 +0900
+++ linux-5.10-rc2.new/lib/fonts/font_8x8.c	2020-11-07 19:19:52.694869415 +0900
@@ -2580,4 +2580,5 @@ const struct font_desc font_vga_8x8 = {
 	.height	= 8,
 	.data	= fontdata_8x8.data,
 	.pref	= 0,
+	.charcount = 255,
 };
diff -uprN linux-5.10-rc2/lib/fonts/font_acorn_8x8.c linux-5.10-rc2.new/lib/fonts/font_acorn_8x8.c
--- linux-5.10-rc2/lib/fonts/font_acorn_8x8.c	2020-11-02 07:43:51.000000000 +0900
+++ linux-5.10-rc2.new/lib/fonts/font_acorn_8x8.c	2020-11-07 19:19:52.694869415 +0900
@@ -276,4 +276,5 @@ const struct font_desc font_acorn_8x8 =
 #else
 	.pref	= 0,
 #endif
+	.charcount = 255,
 };
diff -uprN linux-5.10-rc2/lib/fonts/font_mini_4x6.c linux-5.10-rc2.new/lib/fonts/font_mini_4x6.c
--- linux-5.10-rc2/lib/fonts/font_mini_4x6.c	2020-11-02 07:43:51.000000000 +0900
+++ linux-5.10-rc2.new/lib/fonts/font_mini_4x6.c	2020-11-07 19:19:52.698202820 +0900
@@ -2154,5 +2154,6 @@ const struct font_desc font_mini_4x6 = {
 	.height	= 6,
 	.data	= fontdata_mini_4x6.data,
 	.pref	= 3,
+	.charcount = 255,
 };
 
diff -uprN linux-5.10-rc2/lib/fonts/font_pearl_8x8.c linux-5.10-rc2.new/lib/fonts/font_pearl_8x8.c
--- linux-5.10-rc2/lib/fonts/font_pearl_8x8.c	2020-11-02 07:43:51.000000000 +0900
+++ linux-5.10-rc2.new/lib/fonts/font_pearl_8x8.c	2020-11-07 19:19:52.698202820 +0900
@@ -2584,4 +2584,5 @@ const struct font_desc font_pearl_8x8 =
 	.height	= 8,
 	.data	= fontdata_pearl8x8.data,
 	.pref	= 2,
+	.charcount = 255,
 };
diff -uprN linux-5.10-rc2/lib/fonts/font_sun12x22.c linux-5.10-rc2.new/lib/fonts/font_sun12x22.c
--- linux-5.10-rc2/lib/fonts/font_sun12x22.c	2020-11-02 07:43:51.000000000 +0900
+++ linux-5.10-rc2.new/lib/fonts/font_sun12x22.c	2020-11-07 19:19:52.698202820 +0900
@@ -6162,4 +6162,5 @@ const struct font_desc font_sun_12x22 =
 #else
 	.pref	= -1,
 #endif
+	.charcount = 255,
 };
diff -uprN linux-5.10-rc2/lib/fonts/font_sun8x16.c linux-5.10-rc2.new/lib/fonts/font_sun8x16.c
--- linux-5.10-rc2/lib/fonts/font_sun8x16.c	2020-11-02 07:43:51.000000000 +0900
+++ linux-5.10-rc2.new/lib/fonts/font_sun8x16.c	2020-11-07 19:19:52.708203034 +0900
@@ -274,4 +274,5 @@ const struct font_desc font_sun_8x16 = {
 #else
 	.pref	= -1,
 #endif
+	.charcount = 255,
 };
diff -uprN linux-5.10-rc2/lib/fonts/fonts.c linux-5.10-rc2.new/lib/fonts/fonts.c
--- linux-5.10-rc2/lib/fonts/fonts.c	2020-11-02 07:43:51.000000000 +0900
+++ linux-5.10-rc2.new/lib/fonts/fonts.c	2020-11-07 19:19:52.708203034 +0900
@@ -60,6 +60,10 @@ static const struct font_desc *fonts[] =
 #ifdef CONFIG_FONT_6x8
 	&font_6x8,
 #endif
+#ifdef CONFIG_FONT_16x16_CJK
+#undef NO_FONTS
+    &font_16x16_cjk,
+#endif
 };
 
 #define num_fonts ARRAY_SIZE(fonts)
